src/components/Deck.js

```diff
@@ -18,65 +19,99 @@ import { useTransition, animated } from 'react-spring';
  * essentially it skips animations.
  */
 
-const initialState = { currentSlide: 0, immediate: false };
+const initialState = {
+  currentSlide: 0,
+  immediate: false,
+  immediateElement: false,
+  currentSlideElement: 0,
+  reverseDirection: false
+};
+
+const defaultSlideEffect = {
+  from: {
+    width: '100%',
+    position: 'absolute',
+    transform: 'translate(100%, 0%)'
+  },
+  enter: {
+    width: '100%',
+    position: 'absolute',
+    transform: 'translate(0, 0%)'
+  },
+  leave: {
+    width: '100%',
+    position: 'absolute',
+    transform: 'translate(-100%, 0%)'
+  },
+  config: { precision: 0 }
+};
 
 const Deck = ({ children, loop, keyboardControls, ...rest }) => {
   // Our default effect for transitioning between slides
-  const defaultSlideEffect = {
-    from: {
-      width: '100%',
-      position: 'absolute',
-      transform: 'translate(100%, 0%)'
-    },
-    enter: {
-      width: '100%',
-      position: 'absolute',
-      transform: 'translate(0, 0%)'
-    },
-    leave: {
-      width: '100%',
-      position: 'absolute',
-      transform: 'translate(-100%, 0%)'
-    },
-    config: { precision: 0 }
-  };
+  const { signal, synchronize } = React.useContext(AnimationMutexContext);
+  const [firstRun, setFirstRun] = React.useState(true);
+
   // Check for slides and then number slides.
   const filteredChildren = Array.isArray(children)
-    ? children
-        // filter if is a Slide
-        .filter(x => isComponentType(x, 'Slide'))
+    ? children.filter(x => isComponentType(x, 'Slide'))
     : console.error('No children passed') || [];
 
-  // return a wrapped slide with the animated.div + style prop curried
-  // and a slideNum prop based on iterator
-
-  const Slides = filteredChildren.map((
-    x,
-    i // eslint-disable-next-line react/display-name
-  ) => ({ style }) => (
-    <animated.div style={{ ...style }}>
-      {{
-        ...x,
-        props: { ...x.props, slideNum: i, keyboardControls }
-      }}
-    </animated.div>
-  ));
+  const slideElementMap = React.useMemo(() => {
```

Comment:
We use an element map because for going backwards we need to know how many elements there are to remove them in a backwards-fashion. This also lets us fix bounds issues too.